final-project/application.py

```diff
@@ -0,0 +1,254 @@
+            updated_row = db.execute("UPDATE users SET hash = :new_password WHERE id = :user_id", user_id = session.get("user_id"), new_password = pwd_context.hash(request.form.get("new_password")) )
+            if updated_row == None:
+                return apology("New password has to be different than current")
+        except RuntimeError:
+            # if error with db.execute, apologize to user
+            return apology("500 Error: We'll fix this. Please try again shortly.")
+
+        # redirect user to login page
+        return redirect(url_for("index"))
+
+    # else if user reached route via GET (as by clicking a link or via redirect)
+    else:
+        return render_template("resetPassword.html")
+
+@app.route("/logout")
+def logout():
+    """Log user out."""
+
+    # forget any user_id
+    session.clear()
+
+    # redirect user to login form
+    return redirect(url_for("login"))
+
+@app.route("/register", methods=["GET", "POST"])
+def register():
+    """Register user."""
+
+    # forget any user_id
+    session.clear()
+
+    # if user reached route via POST (as by submitting a form via POST)
+    if request.method == "POST":
+
+        # ensure username was submitted
+        if not request.form.get("username"):
+            return apology("must provide username")
+
+        # ensure password was submitted
+        elif not request.form.get("password"):
+            return apology("must provide password")
+
+        # ensure password_confirmation was submitted
+        elif not request.form.get("password_confirmation"):
+            return apology("passwords did not match")
+
+        try:
+            # create row in database for username
+            rows = db.execute("INSERT INTO 'users' ('id','username','hash') VALUES (NULL, :username, :password)", username=request.form.get("username"), password=pwd_context.hash(request.form.get("password")))
+            if rows== None:
+                return apology("username already exists")
+        except RuntimeError:
+            # if error with db.execute, apologize to user
+            return apology("Error: We'll fix this. Please try again shortly.")
+
+
+        # redirect user to login page
+        return redirect(url_for("login"))
+
+    # else if user reached route via GET (as by clicking a link or via redirect)
+    else:
+        return render_template("register.html")
+
+    return apology("TODO")
+
+
+# TODO: re-write search function to search for products
+@app.route("/search")
+def search():
+    """Search for products that match query."""
+
+    # Query used to create places table
+    # TODO
+
+    # get name of place (postal code or city/town name accepted)
+    q = request.args.get("q") + "%"
+
+    # find matches for the location in search
+    product_search = db.execute("SELECT * FROM todo WHERE todo LIKE :q OR todo LIKE :q LIMIT 50", q=q)
+
+
+    # return JSON array of objects
+    return jsonify(product_search)
+
+
+# TODO: re-write update function
```

Comment:
Even if you need to re-write methods, it's better practice to not leave in commented out code. You can still leave the TODO, and either leave the code live or delete it completely.

Reply:
Good to know. I got lazy since I am the only one working on the project, but it's best to form good habits now. I will go through and remove the old code.

Reply:
🏁 done